share/lib/python/neuron/rxd/region.py

```diff
@@ -396,35 +396,93 @@ def _do_init(self):
             self._xs, self._ys, self._zs = zip(*self._points)
+            maps = {seg:i for i,seg in enumerate([seg for sec in self._secs3d for seg in sec])}
             segs = []
 
             for i, p in enumerate(self._points):
                 if p in surface_voxels:
                     vx = surface_voxels[p]
-                    self._vol[i], self._sa[i], seg = vx[0], vx[1], vx[2]
-
-                    segs.append(seg)
+                    self._vol[i], self._sa[i], idx = vx[0], vx[1], maps[vx[2]]
+                    segs.append(idx)
 
-                    surface_nodes_by_seg.setdefault(seg, [])
-                    surface_nodes_by_seg[seg].append(i)
+                    surface_nodes_by_seg.setdefault(idx, [])
+                    surface_nodes_by_seg[idx].append(i)
 
-                    nodes_by_seg.setdefault(seg, [])
-                    nodes_by_seg[seg].append(i)
+                    nodes_by_seg.setdefault(idx, [])
+                    nodes_by_seg[idx].append(i)
                     self._sa[i] = surface_voxels[p][1]
 
                 else:
                     vx = internal_voxels[p]
-                    self._vol[i], seg = vx[0], vx[1]
-                    segs.append(seg)
+                    self._vol[i], idx = vx[0], maps[vx[1]]
+                    segs.append(idx)
 
-                    nodes_by_seg.setdefault(seg, [])
-                    nodes_by_seg[seg].append(i)
+                    nodes_by_seg.setdefault(idx, [])
+                    nodes_by_seg[idx].append(i)
 
             self._surface_nodes_by_seg = surface_nodes_by_seg
             self._nodes_by_seg = nodes_by_seg 
-            
-            self._segs = list(segs)
-        self._dx = self.dx
+            self._secs3d_names = {sec.hoc_internal_name():sec.nseg for sec in self._secs3d}
+            self._segsidx = segs
+            self._dx = self.dx
+
+    def _segs3d(self, index=None):
+        segs = [seg for sec in self._secs3d for seg in sec]
+        if index:
+            return [seg for i,seg in enumerate(segs) if i in index]
+        return segs
+
+    def _reinit_segs3d(self):
+        """ Assume only nseg has changed, keep the same voxels but change the segment
+            mapping.
+        """
+
+        segs = []
+        nodes_by_seg = {}
+        surface_nodes_by_seg = {}
+        if list(self._secs3d):
+            maps = {seg:i for i,seg in enumerate([seg for sec in self._secs3d for seg in sec])}
+            pts = {}
+            nsegs = {} 
+            for i, sec in enumerate(self._secs3d):
+                rng = range(sec.n3d())
+                pts[sec] = [numpy.array([sec.x3d(i) for i in rng]),
+                            numpy.array([sec.y3d(i) for i in rng]),
+                            numpy.array([sec.z3d(i) for i in rng])]
+                nsegs[sec] = [int(sec.nseg*sec.arc3d(i)/sec.L - 0.5) for i in rng]
+ 
+            def in_seg(i,j,k):
```

Comment:
I'm not sure, but I think is an old rule that we don't use anymore that maps voxels to the closest centroid... but consider the problem of a spine connected to the centroid of a wide dendrite... the closest centroid might be that of the spine even if we're inside the dendrite, so if we plotted the dendrite voxels, there'd be a hole where the spine attaches.

The current rule is to map to whatever segment is closest to the root... so that way the dendrite is intact and spines come out of it.

It would be suboptimal in terms of performance to what you're doing here, but right now, the voxelization algorithm is tied to the nseg discretization, so maybe just re-voxelizing is the way to ensure consistency.